hal/src/thumbv7em/usb/bus.rs

```diff
@@ -754,27 +754,32 @@ impl Inner {
         let mut ep_in_complete = 0;
         let mut ep_setup = 0;
 
+        let intbits = self.usb().epintsmry.read().bits();
+
         for ep in 0..8u16 {
             let mask = 1 << ep;
 
             let idx = ep as usize;
 
-            if let Ok(bank1) = self.bank1(EndpointAddress::from_parts(idx, UsbDirection::In)) {
-                if bank1.is_transfer_complete() {
-                    bank1.clear_transfer_complete();
-                    ep_in_complete |= mask;
-                    // Continuing (and hence not setting masks to indicate complete
-                    // OUT transfers) is necessary for operation to proceed beyond
-                    // the device-address + descriptor stage. The authors suspect a
-                    // deadlock caused by waiting on a write when handling a read
-                    // somewhere in an underlying class or control crate, but we
-                    // can't be sure. Either way, if a write has finished, we only
-                    // set the flag for a completed write on that endpoint index.
-                    // Future polls will handle the reads.
-                    continue;
+            if (intbits & mask) != 0 {
```

Comment:
This is a surprisingly significant optimisation - I've set up a very basic test with a SAMD21 firmware that has 3 USB classes, and at idle it was spending about 3.8% of time in the USB ISR, but with this it's down to 2.6%.  I think it might be worth adding a field to `Inner` that tracks whether any OUT endpoints are ready, then reinstating some of the short-circuits that this PR currently removes.  Will do some tests soon.

Reply:
Sorry, I've not gotten around to doing this yet.  In the spirit of "make it work, then make it fast", I think we're better off leaving optimisations for a different PR.